### PR TITLE
[WIP - Don't merge] Update IoT sdk to version 1.20.3

### PR DESCRIPTION
**Pull request by fbeltrao**
**Changes**

- Add error logs to Azure Function if called with incompatible version
- Add DeviceClient abstraction (IIoTHubDeviceClient) enabling unit testing
- Updated following packages:
  - Microsoft.Azure.Devices: 1.17.0 -> 1.18.1
  - StackExchange.Redis: 2.0.513,2.0.519 -> 2.0.601
  - Microsoft.Azure.Devices.Client: 1.17.0,1.18.1 -> 1.20.3
- Receiving a cloud to device message has been refactored due to DeviceClient.ReceiveAsync not respecting timeout and cancellation tokens. Calls won't start a new request if a pending task exists. Timeout specified by caller is respected.
- Class-C devices end to end tests have been enabled (was disabled due to problem with edgeHub 1.0.6)
- Using edge 1.0.7 in CI

### CODE REVIEW

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs [hunk 1]
```diff
@@ -155,12 +154,165 @@ public async Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary
             return false;
         }
 
+        readonly SemaphoreSlim receiveAsyncLock = new SemaphoreSlim(1, 1);
```
Comment (debugerr) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: nit: can you make those explicitly private and move them to the other declarations?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs [hunk 2]
```diff
@@ -155,12 +154,165 @@ public async Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary
...
+        /// </summary>
+        private TaskCompletionSource<Message> SetupNewReceiveAsyncTaskCompletionSource(TimeSpan timeout)
+        {
+            var localPendingReceiveAsync = this.receiveAsyncTaskCompletionSource = new TaskCompletionSource<Message>();
+            this.receiveAsyncTaskWaitCount = 0;
+
+            _ = this.InternalReceiveAsync(timeout).ContinueWith(async (t) =>
+            {
+                // TODO: remove before PR!
+                Logger.Log(this.devEUI, $"starting new ReceiveAsync task", LogLevel.Debug);
```
Comment (debugerr) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: remove

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs [hunk 3]
```diff
@@ -155,12 +154,165 @@ public async Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary
...
+                    // TODO: remove before PR!
+                    Logger.Log(this.devEUI, $"task ReceiveAsync found message but not one is awaiting, abandoning", LogLevel.Debug);
+
+                    try
+                    {
+                        await this.AbandonAsync(t.Result);
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.Log(this.devEUI, $"failed to abandon message from task without listener: {ex.Message}", LogLevel.Debug);
```
Comment (debugerr) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: shouldn't we Log an error here?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs [hunk 4]
```diff
@@ -155,12 +154,165 @@ public async Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary
...
+                    // Cancel the timer tasks so that it does not fire
+                    cts.Cancel();
+
+                    Task<Message> singleFinished;
+                    await this.receiveAsyncLock.WaitAsync();
+                    try
+                    {
+                        if (localPendingReceiveAsync == this.receiveAsyncTaskCompletionSource)
+                        {
+                            Logger.Log(this.devEUI, $"task ReceiveAsync returned before timeout", LogLevel.Debug);
```
Comment (debugerr) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: Do you want to keep that?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs [hunk 5]
```diff
@@ -155,12 +154,165 @@ public async Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary
...
+                    if (singleFinished != null && singleFinished.IsFaulted)
+                    {
+                        Logger.Log(this.devEUI, $"error in task checking cloud to device message: {singleFinished.Exception?.Message}", LogLevel.Error);
+                        return null;
+                    }
+
+                    return singleFinished?.Result;
+                }
+                else
+                {
```
Comment (debugerr) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: nit: can we add a comment here that this is where we lost the race/timed out, and stop listening for callbacks

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs (hunk 1 above)
Comment (debugerr) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: SemaphoreSlim has a Dispose and that should be called to cleanup unmanaged resources and avoid finalization.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs [hunk 6]
```diff
@@ -155,12 +154,165 @@ public async Task<bool> SendEventAsync(LoRaDeviceTelemetry telemetry, Dictionary
...
+                        localPendingReceiveAsync.SetResult(t.Result);
+                    }
+                    else
+                    {
+                        localPendingReceiveAsync.SetException(t.Exception);
+                    }
+                }
+                else if (t.IsCompletedSuccessfully && t.Result != null)
+                {
+                    // TODO: remove before PR!
```
Comment (TechPreacher) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: One more "remove before PR".

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs (hunk 2 above)
Comment (fbeltrao) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: I will leave verbose log until release 1.0.3

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs (hunk 4 above)
Comment (fbeltrao) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: I will leave verbose log until release 1.0.3

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs (hunk 6 above)
Comment (fbeltrao) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceClient.cs]: I will leave verbose log until release 1.0.3

**Comment by Mandur** [comment]
The IoT hub SDK was updated as part of a different workstream, closing this PR for now

**Pull request closed without merging**